mkosi/manifest.py

```diff
@@ -0,0 +1,130 @@
+from subprocess import DEVNULL, PIPE
+from textwrap import dedent
+
+from typing import cast, Any, Dict, IO, List, Optional
+
+from .backend import run, CommandLineArguments, PackageType
+
+
+@dataclasses.dataclass
+class PackageManifest:
+    """A description of a package
+
+    The fields used here must match
+    https://systemd.io/COREDUMP_PACKAGE_METADATA/#well-known-keys.
+    """
+
+    type: str
+    name: str
+    version: str
+    size: int
+
+    def json(self) -> Dict[str, str]:
+        return {
+            "type": self.type,
+            "name": self.name,
+            "version": self.version,
+        }
+
+
+@dataclasses.dataclass
+class SourcePackageManifest:
+    name: str
+    changelog: Optional[str]
+    packages: List[PackageManifest] = dataclasses.field(default_factory=list)
+
+    def add(self, package: PackageManifest) -> None:
+        self.packages.append(package)
+
+    def report(self) -> str:
+        size = sum(p.size for p in self.packages)
+
+        t = dedent(
+            f"""\
+            SourcePackage: {self.name}
+            Packages:      {" ".join(p.name for p in self.packages)}
+            Size:          {size}
+            """
+        )
+        if self.changelog:
+            t += f"""\nChangelog:\n{self.changelog}\n"""
+        return t
+
+
+@dataclasses.dataclass
+class Manifest:
+    args: CommandLineArguments
+    packages: List[PackageManifest] = dataclasses.field(default_factory=list)
+    source_packages: Dict[str, SourcePackageManifest] = dataclasses.field(default_factory=dict)
+
+    def record_packages(self, root: str) -> None:
+        if cast(Any, self.args.distribution).package_type == PackageType.rpm:
+            self.record_rpm_packages(root)
+        # TODO: add implementations for other package managers
+
+    def record_rpm_packages(self, root: str) -> None:
+        c = run(
+            ["rpm", f"--root={root}", "-qa", "--qf", r"%{NEVRA}\t%{SOURCERPM}\t%{NAME}\t%{SIZE}\n"],
+            stdout=PIPE,
+            stderr=DEVNULL,
+            universal_newlines=True,
+        )
+
+        packages = sorted(c.stdout.splitlines())
+
+        for package in packages:
+            nevra, srpm, name, size = package.split("\t")
+
+            assert nevra.startswith(f"{name}-")
+            evra = nevra[len(name) + 1 :]
+
+            size = int(size)
+
+            try:
+                srpm = self.source_packages[srpm]
+            except KeyError:
+                c = run(
+                    ["rpm", f"--root={root}", "-q", "--changelog", nevra],
+                    stdout=PIPE,
+                    stderr=DEVNULL,
+                    universal_newlines=True,
+                )
+                changelog = c.stdout.strip()
+                srpm = SourcePackageManifest(srpm, changelog)
+                self.source_packages[srpm.name] = srpm
+
+            package = PackageManifest("rpm", name, evra, size)
+            self.packages.append(package)
+            srpm.add(package)
```

Comment:
I don't have the numbers to back this up, but Python exceptions are built around the assumption, that triggering them is rare, because the stack unwinding is expensive, but this looks like it will be more common, so a check `if srpm in self.source_packages` might be a bit faster. Since I don't have a RPM system lying around I can't readily test the difference.

Reply:
Indeed the version with `if` is faster: the try..except block costs a bit, and dict lookup is O(1) (and I think the most recent lookup is also cached specially, so it is a very small O(1)). I'll convert that.